### ISSUE TITLE
[proposal] Add Migrate error trapping or onError method for clearing artefacts when a migration fails

### ISSUE THREAD

**Issue by jammy-git** [laravel/ideas#2089]
This morning I tried to run the following migration:

```
<?php

use Illuminate\Database\Migrations\Migration;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;

class CreateGroupsTable extends Migration
{
    /**
     * Run the migrations.
     *
     * @return void
     */
    public function up()
    {
        Schema::create('groups', function (Blueprint $table) {
            $table->increments('id');
            $table->string('name');
            $table->timestamps();
        });

        Schema::table('users', function(Blueprint $table) {
            $table->unsignedBigInteger('group_id')->after('id');
        });

        Schema::table('users', function(Blueprint $table) {
            $table->foreign('group_id')->references('id')->on('groups');
        });
    }

    /**
     * Reverse the migrations.
     *
     * @return void
     */
    public function down()
    {
        Schema::table('users', function (Blueprint $table) {
            $table->dropForeign(['group_id']);
        });

        Schema::table('users', function (Blueprint $table) {
            $table->dropColumn(['group_id']);
        });

        Schema::dropIfExists('groups');
    }
}
```

*Problem:*
As you might have noticed, I got the table name wrong when creating the foreign key, so the migration failed, though not before creating the `groups` table. The problems then come in if I try to re-run the migration, I get an error that the table already exists. If I try to rollback, because the migration didn't complete, it doesn't try to rollback this migration. So I have no option but to log in to the database directly in order to manually delete the database.

*Solution:*
I propose that we can define an `error` method in which we can specify any actions to run if an error occurs during the migration:

```
  /** 
   * Clear up any database artefacts on error
   *
   * @return void
   */
  public function error()
  {
      Schema::dropIfExists('groups');

      // or

      $this->down();
  }
```

At least this would give us MySQL users some way of specifying what to do when a specific migration fails, since we can't use transactions.

**Comment by LIQRGV** [comment]
If I may suggest, it's better to doing rollback right away.
So we don't have to add 1 more `error` block.

**Comment by jammy-git** [comment]
> If I may suggest, it's better to doing rollback right away.
> So we don't have to add 1 more `error` block.

A rollback will not include a failed migration because they don't get added to the migrations table.

**Comment by LIQRGV** [comment]
No, not the rollback via `php artisan migrate:rollback` or such.
But, it's something like `up` is writen inside a transaction.
```
    public function up()
    {
        DB::transaction(function($context) {
            // the rest of your migration
        })
    }

```
this way, if something wrong happen, all previous changes is reverted.
Because we don't want to add that to migration, we will add that to laravel's migration handler.
Idk which class / file is it.

But ya, we don't have to add something to migrations table.

**Comment by sisve** [comment]
Migrations are already executed within a transaction for those databases that supports it. This is currently only SQL Server and postgresql. If mysql supports this, perhaps in a specific version, modify the MySqlGrammar to enable it in those scenarios.

Note that `$migration->withinTransaction` defaults to true.

        $this->getSchemaGrammar($connection)->supportsSchemaTransactions()
            && $migration->withinTransaction
                    ? $connection->transaction($callback)
                    : $callback();

https://github.com/laravel/framework/blob/40fd590b93f060f2ebe407b3e6d29e66b0e57fbb/src/Illuminate/Database/Migrations/Migrator.php#L398-L401

**Comment by jammy-git** [comment]
Yes, as I alluded to in my proposal, this would really only concern those of us using MySQL (and not InnoDB).

**Comment by LIQRGV** [comment]
Ah I see... 
So it should be MySqlGrammar that changed instead of migration file.
